tools/indexer/cmd/indexer/main.go

```diff
@@ -0,0 +1,434 @@
+		Value:   512,
+	})
+
+	optionBackfillBatch = altsrc.NewIntFlag(&cli.IntFlag{
+		Name:    "backfill-batch",
+		Usage:   "batch size for backfill operations",
+		EnvVars: []string{"INDEXER_BACKFILL_BATCH"},
+		Value:   50,
+	})
+
+	optionHTTPTimeout = altsrc.NewDurationFlag(&cli.DurationFlag{
+		Name:    "http-timeout",
+		Usage:   "HTTP client timeout",
+		EnvVars: []string{"INDEXER_HTTP_TIMEOUT"},
+		Value:   15 * time.Second,
+	})
+)
+
+func createOptionsFromCLI(c *cli.Context) *config.Config {
+	return &config.Config{
+		BlockTick:        c.Duration("block-interval"),
+		ValidatorWait:    c.Duration("validator-delay"),
+		BackfillEvery:    c.Duration("backfill-every"),
+		BackfillLookback: int64(c.Int("backfill-lookback-slots")),
+		BackfillBatch:    c.Int("backfill-batch"),
+		HTTPTimeout:      c.Duration("http-timeout"),
+		OptInContract:    c.String("opt-in-contract"),
+		EtherscanKey:     c.String("etherscan-api-key"),
+		InfuraRPC:        c.String("infura-rpc"),
+		BeaconBase:       c.String("beacon-base"),
+	}
+}
+
+func startIndexer(c *cli.Context) error {
+
+	dbURL := c.String(optionDatabaseURL.Name)
+	infuraRPC := c.String(optionInfuraRPC.Name)
+	beaconBase := c.String(optionBeaconBase.Name)
+	// Initialize random seed
+	rand.Seed(time.Now().UnixNano())
+
+	// Load configuration
+
+	// Validate required configuration
+
+	log.Printf("[INIT] Starting blockchain indexer with StarRocks database")
+	log.Printf("[CONFIG] Block interval: %s, Validator delay: %s, Backfill every: %s",
+		c.Duration("block-interval"), c.Duration("validator-delay"), c.Duration("backfill-every"))
+
+	// Setup graceful shutdown
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer cancel()
+
+	// Connect to StarRocks database
+	db, err := database.MustConnect(ctx, dbURL, 20, 5)
+	if err != nil {
+		log.Fatalf("[DB] Connection failed: %v", err)
+	}
+	defer db.Close()
+	log.Printf("[DB] Connected to StarRocks database")
+
+	// Ensure required tables exist
+	if err := db.EnsureStateTable(ctx); err != nil {
+		log.Fatalf("[DB] Failed to ensure state table: %v", err)
+	}
+	log.Printf(" [DB] State table ready")
+
+	// Initialize HTTP client
+	httpc := httputil.NewHTTPClient(c.Duration("http-timeout"))
+	log.Printf("[HTTP] Client initialized with %s timeout", c.Duration("http-timeout"))
+
+	// Load relay configurations
+	relays, err := relay.UpsertRelaysAndLoad(ctx, db)
+	if err != nil {
+		log.Fatalf("[RELAYS] Failed to load: %v", err)
+	}
+	log.Printf("[RELAYS] Loaded %d active relays:", len(relays))
+	for _, r := range relays {
+		log.Printf("  - Relay ID %d: %s", r.ID, r.URL)
+	}
+
+	// Initialize starting block number
+	lastBN, found := db.LoadLastBlockNumber(ctx)
+	if !found || lastBN == 0 {
+		log.Printf(" [INIT] No previous state found, checking database for latest block...")
+		err := db.Conn.QueryRowContext(ctx, `SELECT COALESCE(MAX(block_number),0) FROM blocks`).Scan(&lastBN)
```

Comment:
If the point of the DB is to mask the underlying storage layer, we shouldnt be accessing the Conn directly. Also if the DB layer is changed, for eg a new table instead of blocks is added, this will also have to be changed. The Conn should be unexported. Any functionality should be provided with the help of functions.